### ISSUE TITLE
High CPU usage after upgrade from 1.12 to 1.18

### ISSUE THREAD

**Issue by Bschitter** [pgbouncer/pgbouncer#844]
We have recently upgraded from pgbouncer 1.12 to pgbouncer 1.18 (and PostgreSQL 11 to 15) and now we see our pgbouncer at nearly 100% CPU (of 1 Core) all the time. Before it was around 20%. The only other thing that we changed was `auth_type` from `md5` to `scram-sha-256`

We currently seem to have a bottleneck in pgbouncer.

Is SCRAM using this much more CPU?
Is this even "calculated" in the pgbouncer process or in the postgres process?
Is there any other reason for this high CPU usage?

**Comment by Bschitter** [comment]
Ok, I will try it.
We have many short lived connections and therefore many authentications. So I would not be surprised...

TLS is off.

**Comment by Bschitter** [comment]
A question in the meantime:
Does it make a difference (performance wise) if the SCRAM passwords in `userlist.txt` are in plain-text or in SCRAM format?

**Comment by JelteF** [comment]
I'm not sure. I haven't done benchmarking of SCRAM authentication. But I would be quite surprised if it mattered. 

To be clear, how is your auth setup? Is the client connecting to pgbouncer with SCRAM, or pgbouncer to postgres with SCRAM, or both?

**Comment by Bschitter** [comment]
I assume it is both.

I have set `auth_type = scram-sha-256` in the `pgbouncer.ini` and `password_encryption = scram-sha-256` (default) in `postgresql.conf`

**Comment by JelteF** [comment]
If you're running pgbouncer on the same server as your application, you might not need scram between your client and pgbouncer. For that part you can probably use Unix sockets and use `peer` or `trust` authentication.

**Comment by JelteF** [comment]
Also, I'd recommend you use the newly released 1.19. It fixes a bug in 1.18 that quite some people ran into.

**Comment by Bschitter** [comment]
We do not run pgbouncer on the application server, we are actually running it on the database server.

I have read about the important bug fix, but I could not see a relation between the changelog and our problem.
Could you please link the github issue?

**Comment by JelteF** [comment]
> We do not run pgbouncer on the application server, we are actually running it on the database server.

Then the opposite optimization is probably possible. i.e. make pgbouncer connect to postgres using unix sockets and use peer/trust auth there. That should at least reduce half the scram code.

> I have read about the important bug fix, but I could not see a relation between the changelog and our problem.
Could you please link the github issue?

It's not related to your perf issue. I meant it more as general advice that it's not recommended to run 1.18.

**Comment by Bschitter** [comment]
Yes, that would be an option, thanks. Although then we would trust all local connections, which I generally prefer not to do.

Ok, thanks for the info. We will plan the upgrade then.

**Comment by Bschitter** [comment]
I performed some mini load tests on our lab system (2 cores) via one of our API which uses the database.
Here are the results:

**1. Response times, n=20**

| auth_type | scram-sha-256 | md5 |
|---------------|---------------|-----|
| **Average**   | 93.7ms            | 62.35ms |
| **Max**       | 151ms             | 74ms    |
| **Min**       | 81ms              | 54ms    |

So it seems we have an additional latency of around ~30ms when using pgbouncer with SCRAM.

**2. Mini load test with [Taurus](https://gettaurus.org/)**

| auth_type      | scram-sha-256 | md5 |
|--------------------|-------------------|---------|
| **requests sent**  |              9305 |   18215 |
| **avg. time (ms)** |               291ms |     148ms |
| **concurrency**    |                20 |      20 |
| **ramp-up**        |               30s |     30s |
| **hold-for**       |                2m |      2m |
| **avg. pgbouncer CPU usage**  |            27.00% |   4.00% |
| **avg.  DB Host CPU usage**    |            48.00% |  50.00% |

So we have around 7x more CPU usage of pgbouncer while the response times approximately double.

We will switch the `auth_type` back from `scram-sha-256` to `md5` on our prod systems.

Thanks for your help!
